Sources/NIOCore/ByteBuffer-binaryEncodedLengthPrefix.swift

```diff
@@ -0,0 +1,337 @@
+    ///  - string: The string to be written.
+    ///  - strategy: The encoding strategy to use.
+    /// - Returns: The total bytes written. This is the bytes needed to write the length, plus the length of the string itself.
+    @discardableResult
+    @inlinable
+    public mutating func writeLengthPrefixedString<
+        Strategy: NIOBinaryIntegerEncodingStrategy
+    >(
+        _ string: String,
+        strategy: Strategy
+    ) -> Int {
+        // writeString always writes the String as UTF8 bytes, without a null-terminator
+        // So the length will be the utf8 count
+        self.reserveCapacity(minimumWritableBytes: string.utf8.count + strategy.requiredBytesHint)
+        var written = 0
+        written += self.writeEncodedInteger(string.utf8.count, strategy: strategy)
+        written += self.writeString(string)
+        return written
+    }
+
+    /// Write the length of `bytes` using `strategy`. Then write the bytes.
+    /// - Parameters:
+    ///  - bytes: The bytes to be written.
+    ///  - strategy: The encoding strategy to use.
+    /// - Returns: The total bytes written. This is the bytes needed to write the length, plus the length of the bytes themselves.
+    @discardableResult
+    @inlinable
+    public mutating func writeLengthPrefixedBytes<
+        Bytes: Sequence,
+        Strategy: NIOBinaryIntegerEncodingStrategy
+    >(
+        _ bytes: Bytes,
+        strategy: Strategy
+    ) -> Int
+    where Bytes.Element == UInt8 {
+        let numberOfBytes = bytes.withContiguousStorageIfAvailable { b in
+            UnsafeRawBufferPointer(b).count
+        }
+        if let numberOfBytes {
+            self.reserveCapacity(minimumWritableBytes: numberOfBytes + strategy.requiredBytesHint)
+            var written = 0
+            written += self.writeEncodedInteger(numberOfBytes, strategy: strategy)
+            written += self.writeBytes(bytes)
+            return written
+        } else {
+            return self.writeLengthPrefixed(strategy: strategy) { buffer in
+                buffer.writeBytes(bytes)
+            }
+        }
+    }
+}
+
+extension ByteBuffer {
+    /// Creates `requiredSpace` bytes of free space immediately before `index`.
+    /// e.g. given [a, b, c, d, e, f, g, h, i, j] and calling this function with (before: 4, requiredSpace: 2) would result in
+    /// [a, b, c, d, 0, 0, e, f, g, h, i, j]
+    /// 2 extra bytes of space were created before index 4 (the letter e).
+    /// The total bytes written will be equal to `requiredSpace`, and the writer index will be moved accordingly.
+    @usableFromInline
+    mutating func _createSpace(before index: Int, requiredSpace: Int) {
+        let bytesToMove = self.writerIndex - index
+
+        // Add the required number of bytes to the end first
+        self.writeRepeatingByte(0, count: requiredSpace)
+        // Move the data forward by that many bytes, to make space at the front
+        self.withVeryUnsafeMutableBytes { pointer in
```

Comment:
could we avoid `*Unsafe*` here?

Reply:
Not really: this is the safe wrapper around the access.

Ultimately this is implementing `memmove` on `ByteBuffer`. Implementing it safely is somewhere between very hard and impossible using the APIs we currently have. We can run through the options:

1. Bytewise copy using `writeInteger`. This is going to be very costly for large moves, which is the concern we have in this method.
2. Manually-autovectorised copy using `writeMultipleInteger` and wider integers. This can work, but it's still going to be _very_ slow compared to `memmove` and it will be much more complex code as we have to handle a wide range of edge-cases.
3. Move using `writeImmutableBuffer`. This will, unfortunately, trigger a CoW.
4. Move using `ByteBufferView`. This will also trigger a CoW.

I fundamentally can't see an implementation here that performs well enough to be used on larger buffers, which is also the circumstance in which it is most likely that we'll need a move. But I'm open to other options.

Reply:
Okay, if this is necessary functionality, can we add it to `ByteBuffer-core.swift` and use in other places too? If my memory serves me right, then @glbrntt added some memory-moving API already a few years back, might be wrong about that though

Reply:
Close enough, I added `copyBytes(at:to:length:)` 5 years ago as part of making `ByteBufferView` mutable. Good memory!

 https://github.com/apple/swift-nio/blame/282f5935cf3352b3d026c35eb57cb3619dd9536f/Sources/NIOCore/ByteBuffer-core.swift#L1077

Reply:
@hamzahrmalik isn't that exactly what we need? `memmove` == copying bytes too

Reply:
Yeah, this looks like we could use it here.

Reply:
Yep, this works perfectly